File: Tugas1_py/src/testing_package/src/commander.py
#!/usr/bin/env python

import rospy
from testing_package.msg import DronePos  # Import the custom message type

def commander_callback(data):
    rospy.loginfo("Received drone status: {}".format(data))
    if(data.x == 1000): 
        rospy.signal_shutdown("Time Out")
    
def commander_node():
    rospy.init_node('commander', anonymous=True)
    rospy.Subscriber('/drone_status', DronePos, commander_callback)
    rospy.spin()

if __name__ == '__main__':
    try:
        commander_node()
    except rospy.ROSInterruptException:
        pass

File: Tugas1_cpp/src/testing_package/src/commander.cpp
#include "ros/ros.h"
#include "testing_package/DronePos.h"
#include "commander.hpp"

void Commander::commanderCallBack(const testing_package::DronePos::ConstPtr &msg) {
  ROS_INFO("Received drone status: %f, %f, %f, %s", msg->x, msg->y, msg->z,
           msg->status.c_str());
}
Commander::Commander():nh(), sub(nh.subscribe("/drone_status", 10, &Commander::commanderCallBack,
                             this)) {}

void Commander::Spin() { ros::spin(); }

int main(int argc, char **argv) {
  ros::init(argc, argv, "commander");
  Commander commander;
  commander.Spin();
  return 0;
}
